tests/memory_tests/test_multiple_runs.py

```diff
@@ -68,28 +81,132 @@ def run_problem():
+
+            # Calculate memory difference for this run
+            run_memory_diff = post_run_memory - pre_run_memory
+            memory_measurements.append((i + 1, run_memory_diff))
+
+            print(f"Memory difference for run {i + 1}: {run_memory_diff:.3f} MiB")
+
+            # Analyze memory changes for this specific run
+            print(f"\nTop memory changes during run {i + 1}:")
+            top_stats = post_snapshot.compare_to(pre_snapshot, "lineno")
+            for j, stat in enumerate(top_stats[:5]):  # Show top 5 changes
+                size_diff = stat.size_diff / 1024 / 1024  # Convert to MiB
+                print(f"  {j + 1}. {size_diff:+.3f} MiB | {stat.traceback.format()[-1].strip()}")
+                if j == 0 and abs(size_diff) > 1.0:  # Show full traceback for largest change
+                    print("     Full traceback:")
+                    for line in stat.traceback.format():
+                        print(f"       {line.strip()}")
+
+            # Check if this run exceeded positive threshold
+            if run_memory_diff > MEMORY_DIFF_THRESHOLD:
+                print(
+                    f"\nWARNING: Run {i + 1} memory difference ({run_memory_diff:.3f} MiB) "
+                    f"exceeds threshold ({MEMORY_DIFF_THRESHOLD} MiB)"
+                )
+        # Final analysis
+        print(f"\n{'=' * 60}")
+        print("MEMORY ANALYSIS SUMMARY")
+        print(f"{'=' * 60}")
+
+        gc.collect()
+
+        final_memory = print_memory_state("Final state (after garbage collection)")
+        final_diff = final_memory - baseline_memory
+
+        print("\nMemory measurements per run:")
+        for run_num, diff in memory_measurements:
+            status = "PASS" if abs(diff) <= MEMORY_DIFF_THRESHOLD else "FAIL"
+            print(f"  Run {run_num}: {diff:+.3f} MiB [{status}]")
+
+        print(f"\nOverall memory increase (after garbage collection): {final_diff:.3f} MiB")
+
+        # Overall memory analysis from start to finish
+        print("\nOverall top memory changes (start to finish):")
+        final_snapshot = tracemalloc.take_snapshot()
+        overall_stats = final_snapshot.compare_to(snapshot_start, "lineno")
+        for i, stat in enumerate(overall_stats[:8]):
+            size_diff = stat.size_diff / 1024 / 1024
+            count_diff = stat.count_diff
+            print(
+                f"  {i + 1}. {size_diff:+.3f} MiB ({count_diff:+d} blocks) | "
+                f"{stat.traceback.format()[-1].strip()}"
+            )
+
+        # Cross-run memory leak detection
+        print("\nMemory leak analysis:")
+        positive_diffs = [
+            diff for _, diff in memory_measurements if diff > 0.1
+        ]  # Only significant positive diffs
+        avg_increase = sum(positive_diffs) / len(positive_diffs)
+        print(f"  Average memory increase per run: {avg_increase:.3f} MiB")
+        if avg_increase > MEMORY_DIFF_THRESHOLD * 0.5:
```

Comment:
So this is in case the increase is big but not big enough to cause a crash ? For debug then ?